### PR TITLE
Complete JQ Implementation

### PR DESCRIPTION
**Pull request by b-wager**
Ok so I did like half the implementation but it's 3 am and I'm tired.

### CODE REVIEW

File: mini-projects/00-jq-rs/Cargo.toml [hunk 1]
```diff
@@ -4,3 +4,9 @@ version = "0.1.0"
 edition = "2021"
 
 [dependencies]
+
+anyhow = "1.0.86"
+clap = { version = "4.5.20", features = ["derive"] }
+colored = "2.1.0"
+regex = "1.10.6"
+serde_json = "1.0.128"
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/Cargo.toml:12]: nit: you could use the `preserve_order` feature to preserve the original order of the json when `sort-keys` isn't specified

File: mini-projects/00-jq-rs/src/filter.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
+use std::str::Chars;
+
+use serde_json::Value;
+
+pub fn filter(json: &Value, filter: &str) -> Value {
+    let mut values = json.clone();
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filter.rs:6]: nit: could have passed a mutable reference to avoid the need to copy

File: mini-projects/00-jq-rs/src/filter.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        } else {
+            let filter = &filter.chars().skip(1).collect::<String>();
+            if values.is_object() {
+                values = values.get(filter).unwrap_or(&Value::Null).clone();
+            } else {
+                panic!("Cannot index into a non-object");
+            }
+        }
+    } else {
+        panic!("Filter must start with a '.'");
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filter.rs:31]: this isn't necessarily true, the built-in functions do not start with a "."

File: mini-projects/00-jq-rs/src/filter.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        } else if filter_chars.next() == Some('[') {
+            values = index_array(json, filter_chars);
+        } else {
+            let filter = &filter.chars().skip(1).collect::<String>();
+            if values.is_object() {
+                values = values.get(filter).unwrap_or(&Value::Null).clone();
+            } else {
+                panic!("Cannot index into a non-object");
+            }
+        }
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filter.rs:29]: as a whole, this a great first pass at being able to parse multiple different filter types, but this logic could definitely have been made more explicit (maybe through a regex pattern search). I can imagine that, with the way it's currently architected, this wouldn't scale well to 10-20 different filter options. Something that could improve this is having an explicit enum to represent all the filters your application supports, and then being able to explicitly check if the enum's conditions for the filter are met

File: mini-projects/00-jq-rs/src/filter.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        // Parse the index as a single value
+        } else {
+            let index = indexer.parse::<usize>().unwrap();
+            return array
+                .get(index)
+                .unwrap_or_else(|| panic!("Index out of range"))
+                .clone();
+        }
+    } else {
+        panic!("Cannot index into a non-array");
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/filter.rs:82]: I would have loved to see less panics and more bubbling up errors through a custom-defined error type. The benefit of the latter is more verbosity in the type of error, and the ability to recover and get nicer logs of what happened

File: mini-projects/00-jq-rs/src/main.rs [hunk 6]
```diff
@@ -1,3 +1,58 @@
...
+        bail!("Cannot use --indent and --compact-output together")
+    }
+    if args.color_output && args.monochrome_output {
+        bail!("Cannot use --color-output and --monochrome-output together")
+    }
+
+    // Set the indent value and compact_output flags
+    let mut compact_output = args.compact_output;
+    let indent = args.indent.map_or(Ok(2), |i| {
+        if i == 0 {
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:29]: nit: would move hardcoded numbers into constant variables

File: mini-projects/00-jq-rs/src/main.rs [hunk 7]
```diff
@@ -1,3 +1,58 @@
...
+            println!();
+        } else {
+            print_json::print_json(values, indent, indent);
+        }
+    } else if compact_output {
+        print_json::print_compact_json_color(values);
+        println!();
+    } else {
+        print_json::print_json_color(values, indent, indent);
+    }
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/main.rs:55]: this is where you might have gotten a lot of utility out of having a generic Print trait for custom printer structs, so rather than needing to define separate functions, you could just have separate traits with individual structs (that can be configured based on their needs). I think that would allow this to scale better to making adding more printing parameters easier

File: mini-projects/00-jq-rs/src/print_json.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+pub fn print_json_color(json: &Value, current_indent: usize, original_indent: usize) {
+    match json {
+        Value::Object(ref map) => {
+            println!("{{");
+            for (key, value) in map {
+                print!(
+                    "{}{}: ",
+                    " ".repeat(current_indent),
+                    format!("\"{}\"", key).color(Color::Blue).bold()
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/print_json.rs:69]: while using the color crate makes the colors (mostly) correct for the default jq colors, it makes it a lot harder for the user to customize them

File: mini-projects/00-jq-rs/src/print_json.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+                    print!(",");
+                }
+            }
+            print!("]");
+        }
+        Value::String(ref s) => print!("{}", format!("\"{}\"", s).color(Color::Green)),
+        Value::Number(ref n) => print!("{}", n),
+        Value::Bool(ref b) => print!("{}", b),
+        Value::Null => print!("{}", "null".color(Color::Black)),
+    }
```
Comment (ayushchakra) [inline mini-projects/00-jq-rs/src/print_json.rs:119]: as a whole, there seems to be a lot of overlapping utility between functions in this file (especially for printing strings, numbers, bools, and null) - could have benefitted from splitting into their own shared function

**Comment by ayushchakra** [comment]
# Rubric For Mini Project 0: JQ

## Submission

### Canvas

1: The link submitted on Canvas is for the pull request containing the
assignment submission.

0: The link submitted on Canvas is not for the pull request containing the
assignment submission.

### Git Branch

1: The branch includes only changes for this assignment.

0: The branch includes commits or changes outside this assignment.

### Git Commit Message

1: All commit messages for the branch are reasonably clear.

0: One or more commit messages for the branch are reasonably clear.

### GitHub Pull Request

1: The pull request is made to the `main` branch of the student's fork of the
course repository.

0: The pull request is made to a branch other than `main` or to a repo that is
not the student's fork of the course repository.

### Points
-1 for unrelated files
**Submission Score: 3/4**

## Features
### Command Line Interface
For each of the required CLI arguments `filter-string`, `path-to-json`, `color-output`, `monochrome-output`, `sort-keys`, `indent`, and `compact-output`

0.5: The argument can be correctly parsed and consumed by the rest of the package.

0: The argument is not correctly parsed and propagated to the rest of the package.

Additionally, the command line parsing correctly raises errors to the user when conflicting arguments are inputted.

1.5: Errors are properly determined and bubbled up to the user.

0: Errors are not propagated back to the user.

5/5

### Supported Filters
For each of the required filters `identity`, `object-identifier index`, `array index`, `array slice`, `pipe`, and `array iterator`

2: The filter is correctly implemented and properly parses all possible JSON inputs.

1: Minor errors exist in the implement of the filter.

0: Major errors exist in the implementation, or the filter was omitted from the package.

Additionally, comprehensive unit tests are required for each feature:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the filter, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the filter.

0: Tests are missing, or lack significant assessment of the filter's performance.

#### Comments
Identity: Correct implementation and unit tests
Object Identifier Index: Correct implementation and unit tests
Array Index: Correct implementation and unit tests
Array Slice: Correct implementation and unit tests
Pipe: Correct Implementation and unit tests
(-4) Array Iterator: Was not implemented and no unit tests

(-2) Not docing points everywhere for this, but as a whole, you test the positive case for each of your tests where you expect current output, but don't test for edge cases or when you expect your filters to fail (ex. when invalid datatypes are given).

### Built-In Functions
For each of the required built-in functions `add`, `length`, `del`

2: The built-in function is correctly implemented and properly processes all possible JSON inputs.

1: Minor errors exist in the implementation of the operator.

0: Major errors exist in the implementation, or the function was omitted from the package.

Additionally, comprehensive unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to exercise a wide varity of use-cases and edge-cases of the function, ensuring that the output is correct and that errors are properly outputted.

1: Unit tests are present, but lack a significant breadth of cases to fully exercise the function.

0: Tests are missing, or lack significant assessment of the function's performance.

#### Comments
(-12) No built-in function was implemented

### Formatted Printing
For each of the required formats, `colored printing`, `monochrome printing`, `sorted keys`, `indents`, and `compact output`

2: The formatter is correctly implemented and properly displays any JSON object..

1: Minor errors exist in the implementation of the formatter.

0: Major errors exist in the implementation, or the formatter was omitted from the package.

Additionally, unit tests are required for each function:

2: Full unit test coverage is provided (with passing implementations) to ensure that the formatter can correctly handle all expected inputs.

1: Unit tests are present, but lack significant breadth of cases to fully exercise the formatter.

0: Tests are missing, or lack significant assessment of the formatter.

#### Comments
(-1) Colored Printing: By default, color printing works, but you did not use the `JQ_COLORS` env variable, nor the actual VT100 escapes, no unit tests
Monochrome Printing: Correct implementation, no unit tests
Sorted Keys: Serde sorts keys by default (for future reference, you can download a separate version of serde that has the feature of preserving order), no unit tests
Indents: Correctly implemented, no unit tests
Compact Output: Correctly implemented, no unit tests


### Points
**Features Score: 32/61**

## Code Quality

### Design

#### File/Module organization

2: The program is appropriately split into reasonable files/modules

1: The program is somewhat organized, but there's room for improvement (splits aren't idiomatic, or some modules are still too large, etc.)

0: The program is not split at all

#### Scalability

5: The code is architected with scalability in mind. If more features needed to be added to the filtering/printing/CLI, they could easily be integrated.

3: The code was architected with some scalaibility in mind. However, significant refactoring would be needed in order to expand the feature set of this tool. (Possible symptoms include not modularizing enough/not effectively using traits + generics).

0: The code is not scalable. The entire package would need to be refactored in order to expand its features.

#### General Cleanliness

4: The code is effectively documented, especially for sources of user-facing errors. Additionally, function signatures are clear and the code is overall very readable.

2: Attempts at verbose documentation were made, but certain sources of error are unclear OR poor function signatures are used and readers would have slight difficulties while scanning the code.

0: Documentation is severely lacking in several palces and the code is not written in a readable way.

### Lint/Format

#### RustFmt

2: RustFmt produces no warnings when run on the changed files

1: RustFmt produces a few minor warnings when run on the changed files

0: RustFmt produces many warnings when run on the changed files

#### Clippy

2: Clippy produces no warnings when run on the changed files

1: Clippy produces a few minor warnings when run on the changed files

0: Clippy produces many warnings when run on the changed files

### Points
**Code Quality Score: 11/15**
Scalability: 3/5 - as a whole, I like the separation of logic you made between modules of your application. However, I think there are a few more steps you can take (see comments on PR) to better prepare for added features
General Cleanliness:  2/4 - Code could definitely benefit from some added documentation, especially for how you are parsing the filters


## Summary
Add up the point totals here

**Total Score: 46/80**
